### PR TITLE
async'ify msg, don't throw for flow control

### LINKED ISSUE #906: “Receiving end does not exist” error gets thrown and interrupts the dev tools
FF 68.7.0esr
W7
Stylus 1.5.10
http://i.imgur.com/uGX8jQ3.png

#668

It's annoying.

### PR DESCRIPTION
**Pull request by tophf**
Fixes #906.

There should be no functional changes but let's test this for a while anyway.

Code-wise, since conversion to async/await changes the code a lot, I also shuffled and reworked the entire msg.js mainly to reduce indirection i.e. to make call chains shorter hence more obvious while debugging or navigating through the code especially when using "jump to definition" and "peek definition".

### CODE REVIEW

File: js/msg.js [hunk 1]
```diff
@@ -1,258 +1,168 @@
...
+            msg.sendTab(tab.id, data, null, 'both').catch(msg.ignoreError));
         }
-        return Promise.all(requests);
-      });
-  }
+      }
+      return Promise.all(requests);
+    },
 
-  function broadcastExtension(...args) {
```
Comment (eight04) [inline js/msg.js:140]: `broadcastExtension` API is missing.

File: js/msg.js [hunk 2]
```diff
@@ -1,258 +1,168 @@
...
-    return bgReady.then(bg => {
-      if (bg) {
-        if (!bg._msg.handler) {
-          throw new Error('there is no bg handler');
-        }
-        const handlers = bg._msg.handler.extension.concat(bg._msg.handler.both);
+  window.API = new Proxy({}, {
+    get(target, name) {
+      return async (...args) => {
+        const bg = isBg && window || chrome.tabs && (getExtBgIfReady() || await getRuntimeBg());
```
Comment (eight04) [inline js/msg.js:32]: We no longer cache `bg` anymore. Is this intended?

File: js/msg.js [hunk 3]
```diff
@@ -1,258 +1,168 @@
...
-    let result;
-    for (const fn of callbacks) {
-      const data = withPromiseError(fn, ...args);
-      if (data !== undefined && result === undefined) {
-        result = data;
-      }
-    }
-    return result;
+  function getRuntimeBg() {
+    return new Promise(resolve =>
```
Comment (eight04) [inline js/msg.js:133]: I think we can use promisified `browser` API here?

File: js/msg.js [hunk 4]
```diff
@@ -1,258 +1,168 @@
...
-  // {type, error, data, id}
-  function unwrapData(result) {
-    if (result === undefined) {
-      throw new Error('Receiving end does not exist');
-    }
-    if (result.error) {
-      throw Object.assign(new Error(result.data.message), result.data);
-    }
-    return result.data;
+  function unwrapData({data, error} = {}) {
```
Comment (eight04) [inline js/msg.js:163]: This won't throw when the remote handler doesn't exist. It should probably be:
```js
({data, error} = {
  data: {message: 'Receiving end does not exist'},
  error: true
})
```

BTW, maybe we should move the error handling stuff into `polyfill.js` in the future. In Firefox, you can return a promise in `addListener` callback. If it fails, the error object will be passed to the caller correctly. Therefor `msg.js` will only include `sendBg` and the `API` stuff.

File: js/msg.js [hunk 5]
```diff
@@ -1,258 +1,168 @@
...
-    const handlers = message.target === 'tab' ?
-      handler.tab.concat(handler.both) : message.target === 'extension' ?
-      handler.extension.concat(handler.both) :
-      handler.both.concat(handler.extension, handler.tab);
-    if (!handlers.length) {
-      return;
+  function handleMessage({data, target}, sender, sendResponse) {
+    const res = msg._execute(TARGETS[target] || TARGETS.all, data, sender);
+    if (res instanceof Promise) {
+      handleResponseAsync(res, sendResponse);
```
Comment (eight04) [inline js/msg.js:141]: How about renaming it to `wrapData` function (compare to `unwrapData`) and making it returns the message object so we can use it like:
```js
wrapData(res).then(sendResponse);
```

**Review by eight04 — CHANGES_REQUESTED** [review]
👍 the message handler looks a lot cleaner.

**Comment by tophf** [comment]
Thanks for the review!

**Pull request merged**